src/WebJobs.Extensions.DurableTask/Listener/TaskOrchestrationShim.cs

```diff
@@ -234,11 +248,66 @@ private void TraceAndSendExceptionNotification(string exceptionDetails)
             }
         }
 
+        private OrchestrationInvocationResult GetInvocationResult(object ooprocResult, Exception ex = null)
+        {
+            (JObject resultJObject, string resultJSONString) = this.ParseOOProcResult(ooprocResult);
+
+            this.Config.TraceHelper.ProcessingOutOfProcPayload(
+                functionName: this.Context.FunctionName,
+                taskHub: this.Context.HubName,
+                instanceId: this.Context.InstanceId,
+                details: resultJSONString);
+
+            OrchestrationInvocationResult invocationResult = new OrchestrationInvocationResult()
+            {
+                ReturnValue = ooprocResult,
+                Exception = ex,
+                Json = resultJObject,
+                JsonString = resultJSONString,
+            };
+            return invocationResult;
+        }
+
+        private (JObject, string) ParseOOProcResult(object result)
+        {
+            string jsonString;
+            JObject json = result as JObject;
+            if (json == null)
+            {
+                if (result is string text)
+                {
+                    try
+                    {
+                        jsonString = text;
+                        json = JObject.Parse(text);
+                    }
+                    catch
+                    {
+                        throw new ArgumentException("Out of proc orchestrators must return a valid JSON schema");
+                    }
+                }
+                else
+                {
+                    throw new ArgumentException("The data returned by the out-of-process function execution was not valid json.");
+                }
+            }
+            else // result was a JObject all long, need to assign jsonString
+            {
+                jsonString = json.ToString();
+            }
+
+            return (json, jsonString);
+        }
+
         internal class OrchestrationInvocationResult
         {
             public object ReturnValue { get; set; }
 
             public Exception Exception { get; set; }
+
+            public JObject Json { get; set; }
+
+            public string JsonString { get; set; }

```

Comment:
Tip: I recommend making all these properties read-only (removing the setters) and have it get fully initialized in a constructor. That way we can leverage the compiler to help us guarantee that these last two properties are never null. Immutable types FTW!

Reply:
Of course ❤️ ! I made them part of a constructor, but I couldn't get the project to compile with public-readonly fields: Visual Studio really wanted them to be private if I added the `readonly` modifier, which meant I would need to add some boilerplate by creating getters for each property. 

So instead I went for public fields with private setters, let me know what you think!

Reply:
This is the syntax for read-only properties in C# (you just need to remove `set`):
```csharp
public JObject Json { get; }

public string JsonString { get; }
```

This is even better than private setters since not even internal code can modify them later.

Reply:
Oh I see, my mistake was to additionally add the `readonly` keyword. I'll refactor this now

Reply:
You may want to do some final validation here end-to-end. I don't think Newtonsoft can deserialize fields that have no setter method...

Reply:
[This](https://stackoverflow.com/questions/4066947/private-setters-in-json-net) may be the best guidance. If we could use C# 9, you would want init-setters, but I don't think we can get away with that currently.

Maybe a combination of read-only properties with a constructor with the correct parameter names would work?

Reply:
Are these objects ever deserialized? It looked to me like they're only created in code.

Reply:
Ah good call, I was mixing this object up with `OutOfProcOrchestratorState`